File: agent-framework/server/services/llm_logger.py
import time
from typing import Any


class LLMLogger:
    """内存 LLM 请求/响应日志器，按 thread_id 存储每轮 LLM 调用的完整信息"""

    def __init__(self, max_calls_per_thread: int = 50):
        self._storage: dict[str, list[dict]] = {}
        self._max = max_calls_per_thread

    def log_call(
        self,
        thread_id: str,
        request_info: dict,
        response_info: dict,
    ) -> str:
        """记录一次 LLM 调用，返回 call_id"""
        if thread_id not in self._storage:
            self._storage[thread_id] = []

        call_id = f"call-{int(time.time() * 1000)}-{len(self._storage[thread_id])}"
        entry = {
            "call_id": call_id,
            "timestamp": time.time(),
            "request": request_info,
            "response": response_info,
        }
        self._storage[thread_id].append(entry)

        if len(self._storage[thread_id]) > self._max:
            self._storage[thread_id] = self._storage[thread_id][-self._max:]

        return call_id

    def get_calls(self, thread_id: str) -> list[dict]:
        """获取某 thread 的所有 LLM 调用记录"""
        return self._storage.get(thread_id, [])

    def clear_thread(self, thread_id: str):
        """清除某 thread 的所有日志"""
        self._storage.pop(thread_id, None)

    @property
    def thread_count(self) -> int:
        return len(self._storage)

File: agent-framework/src/main/java/io/agentmanager/framework/service/LLMLogger.java
package io.agentmanager.framework.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class LLMLogger {
    private static final Logger log = LoggerFactory.getLogger(LLMLogger.class);

    private final int maxCallsPerThread;
    private final Map<String, List<CallRecord>> storage = new ConcurrentHashMap<>();

    public LLMLogger() {
        this(50);
    }

    public LLMLogger(int maxCallsPerThread) {
        this.maxCallsPerThread = maxCallsPerThread;
    }

    public String logCall(String threadId, Map<String, Object> requestInfo, Map<String, Object> responseInfo) {
        var calls = storage.computeIfAbsent(threadId, k -> new ArrayList<>());
        var callId = "call-" + System.currentTimeMillis() + "-" + calls.size();
        var record = new CallRecord(callId, System.currentTimeMillis(), requestInfo, responseInfo);
        calls.add(record);
        if (calls.size() > maxCallsPerThread) {
            calls.remove(0);
        }
        return callId;
    }

    public List<CallRecord> getCalls(String threadId) {
        var exact = storage.get(threadId);
        if (exact != null) {
            return exact;
        }
        // 兼容 agent_state 中带租户前缀的会话变体（如 "__anon__:debug-user:xxx" 与 "debug-user:xxx" 视为同一会话）
        var normalized = normalizeThreadId(threadId);
        if (normalized != null) {
            for (var entry : storage.entrySet()) {
                if (normalized.equals(normalizeThreadId(entry.getKey()))) {
                    return entry.getValue();
                }
            }
        }
        return List.of();
    }

    /** 剥离租户前缀段（冒号分隔），保留末尾会话本体 */
    private static String normalizeThreadId(String threadId) {
        if (threadId == null || threadId.isBlank()) return null;
        var idx = threadId.lastIndexOf(':');
        return idx >= 0 ? threadId.substring(idx + 1) : threadId;
    }

    public void clearThread(String threadId) {
        storage.remove(threadId);
    }

    public record CallRecord(
        String callId, long timestamp,
        Map<String, Object> request,
        Map<String, Object> response
    ) {}
}
